source/link/linker.cpp

```diff
@@ -16,1183 +16,277 @@
+                       uint32_t options) const {
+  auto linkedModule = MakeUnique<Module>();
+
+  std::vector<std::unique_ptr<Module>> modules;
+  modules.reserve(binaries.size());
+
+  uint32_t id_bound = 0u;
+  for (const auto& mod : binaries) {
+    std::unique_ptr<Module> module = BuildModule(impl_->context->target_env, impl_->context->consumer, mod.data(), mod.size());
+    if (module == nullptr)
+      return 1;
+    module->ForEachInst([&id_bound](Instruction* insn){
+        for (auto o = insn->begin(); o != insn->end(); ++o) {
+          if (!spvIsIdType(o->type))
+            continue;
+          o->words[0] += id_bound;
+        }
+    });
+    id_bound += module->IdBound() - 1u;
+    modules.push_back(std::move(module));
   }
-}
-
-static spv_result_t MergeModules(spv_context_t& context,
-                                 std::vector<std::vector<uint32_t>> &binaries,
-                                 size_t num_binaries, uint32_t max_index,
-                                 uint32_t version,
-                                 spv_binary_t& merged_spvbinary,
-                                 bool create_library) {
-  spv_position_t position = {};
 
-  std::vector<uint32_t> indices(num_binaries, 5u);
-
-  std::unordered_map<spv::Id, const char *> exports;
-  std::unordered_map<spv::Id, const char *> imports;
-  std::unordered_map<spv::Id, spv::Id> link_table;
-  std::vector<spv::Id> builtins;
-
-  uint32_t* merged_binary = merged_spvbinary.code;
-  size_t& merged_length = merged_spvbinary.wordCount;
-
-  // Generate new header
-  SetWord(merged_binary, 0u, spv::MagicNumber);
-  SetWord(merged_binary, 1u, version);
-  SetWord(merged_binary, 2u, 0u);
-  SetWord(merged_binary, 3u, max_index);
-  SetWord(merged_binary, 4u, 0u);
-  merged_length = 5u;
-
-  const auto CopySections = [&binaries,&indices,num_binaries,
-                             &merged_binary,&merged_length](const std::function<bool (spv::Op)>& check_section) {
-    for (uint32_t i = 0u; i < num_binaries; ++i) {
-      const uint32_t* const binary = binaries[i].data();
-      const size_t length = binaries[i].size();
-      const uint32_t index = indices[i];
-      uint32_t j = index;
-      uint32_t prev_j = j;
-
-      while (j < length) {
-        uint32_t word_count;
-        const spv::Op opcode = ParseOpcode(binary, j, word_count);
-
-        prev_j = j;
-
-        if (!check_section(opcode))
-          break;
-
-        j += word_count;
+  for (const auto& module : modules)
+    for (const auto& insn : module->capabilities())
+      linkedModule->AddCapability(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
+    for (const auto& insn : module->extensions())
+      linkedModule->AddExtension(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
+    for (const auto& insn : module->ext_inst_imports())
+      linkedModule->AddExtInstImport(MakeUnique<Instruction>(insn));
+
+  // FIXME
```

Comment:
Use TODO(your-name) rather than FIXME. 
Also say what the issue is.   I  guess it would be that it's an error to mix different memory models? Or is there a promotion lattice you can use?  Ok to error out for now(!)